### PR TITLE
Add ServiceWorkers for PWA-use

### PR DESCRIPTION
**Pull request by bwnstck**
# Description
- PWA with Statusbar
Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: client/src/service-worker.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/* eslint-disable no-restricted-globals */
+
+// This service worker can be customized!
+// See https://developers.google.com/web/tools/workbox/modules
+// for the list of available Workbox modules, or add any other
+// code you'd like.
+// You can also remove this file if you'd prefer not to use a
+// service worker, and the Workbox build step will be skipped.
```
Comment (arassaria) [inline client/src/service-worker.js:8]: Necessary comments?

File: client/src/service-worker.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import { StaleWhileRevalidate } from "workbox-strategies";
+import { BackgroundSyncPlugin } from "workbox-background-sync";
+import { NetworkOnly } from "workbox-strategies";
+import { BroadcastUpdatePlugin } from "workbox-broadcast-update";
+clientsClaim();
+
+// Precache all of the assets generated by your build process.
+// Their URLs are injected into the manifest variable below.
+// This variable must be present somewhere in your service worker file,
+// even if you decide not to use precaching. See https://cra.link/PWA
```
Comment (arassaria) [inline client/src/service-worker.js:23]: Again, necessary?

File: client/src/service-worker.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+precacheAndRoute(self.__WB_MANIFEST);
+
+// Set up App Shell-style routing, so that all navigation requests
+// are fulfilled with your index.html shell. Learn more at
+// https://developers.google.com/web/fundamentals/architecture/app-shell
```
Comment (arassaria) [inline client/src/service-worker.js:28]: Necessary?

File: client/src/service-worker.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+const fileExtensionRegexp = new RegExp("/[^/?]+\\.[^/]+$");
+registerRoute(
+  // Return false to exempt requests from being fulfilled by index.html.
```
Comment (arassaria) [inline client/src/service-worker.js:31]: Comment...

File: client/src/service-worker.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  ({ request, url }) => {
+    // If this isn't a navigation, skip.
```
Comment (arassaria) [inline client/src/service-worker.js:33]: And again

File: client/src/service-worker.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    if (request.mode !== "navigate") {
+      return false;
+    } // If this is a URL that starts with /_, skip.
```
Comment (arassaria) [inline client/src/service-worker.js:36]: Why so many comments?

File: client/src/service-worker.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (url.pathname.startsWith("/_")) {
+      return false;
+    } // If this looks like a URL for a resource, because it contains // a file extension, skip.
```
Comment (arassaria) [inline client/src/service-worker.js:40]: Pls stop.

File: client/src/service-worker.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (url.pathname.match(fileExtensionRegexp)) {
+      return false;
+    } // Return true to signal that we want to use the handler.
```
Comment (arassaria) [inline client/src/service-worker.js:44]: Its already been too many

File: client/src/service-worker.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    return true;
+  },
+  createHandlerBoundToURL(process.env.PUBLIC_URL + "/index.html")
+);
+
+// An example runtime caching route for requests that aren't handled by the
+// precache, in this case same-origin .png requests like those from in public/
```
Comment (arassaria) [inline client/src/service-worker.js:52]: I don't know what to write anymore, so i will comment with an elephant now. 🐘

File: client/src/service-worker.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+registerRoute(
+  // Add in any other file extensions or routing criteria as needed.
```
Comment (arassaria) [inline client/src/service-worker.js:54]: 🐘

File: client/src/service-worker.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  ({ url }) =>
+    url.origin === self.location.origin && url.pathname.endsWith(".png"), // Customize this strategy as needed, e.g., by changing to CacheFirst.
+  new StaleWhileRevalidate({
+    cacheName: "images",
+    plugins: [
+      // Ensure that once this runtime cache reaches a maximum size the
+      // least-recently used images are removed.
```
Comment (arassaria) [inline client/src/service-worker.js:61]: 🐘

File: client/src/service-worker.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      new ExpirationPlugin({ maxEntries: 50 }),
+    ],
+  })
+);
+
+// This allows the web app to trigger skipWaiting via
+// registration.waiting.postMessage({type: 'SKIP_WAITING'})
```
Comment (arassaria) [inline client/src/service-worker.js:68]: 🐘

File: client/src/service-worker.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+self.addEventListener("message", (event) => {
+  if (event.data && event.data.type === "SKIP_WAITING") {
+    self.skipWaiting();
+  }
+});
+
+// Any other custom service worker logic can go here.
+
+// https://developers.google.com/web/tools/workbox/modules/workbox-background-sync
```
Comment (arassaria) [inline client/src/service-worker.js:77]: 🐘

File: client/src/service-worker.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+const bgSyncPlugin = new BackgroundSyncPlugin("postQueue", {
+  maxRetentionTime: 24 * 60, // Retry for max of 24 Hours (specified in minutes)
```
Comment (arassaria) [inline client/src/service-worker.js:79]: 🐘

File: client/src/service-worker.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+registerRoute(
+  /\/api\/.*\/*/,
+  new NetworkOnly({
+    plugins: [bgSyncPlugin],
+  }),
+  "POST"
+);
+
+// https://developers.google.com/web/tools/workbox/modules/workbox-broadcast-update
```
Comment (arassaria) [inline client/src/service-worker.js:90]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+// register() is not called by default.
+
+// This lets the app load faster on subsequent visits in production, and gives
+// it offline capabilities. However, it also means that developers (and users)
+// will only see deployed updates on subsequent visits to a page, after all the
+// existing tabs open on the page have been closed, since previously cached
+// resources are updated in the background.
+
+// To learn more about the benefits of this model and instructions on how to
+// opt-in, read https://cra.link/PWA
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:11]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+
+const isLocalhost = Boolean(
+  window.location.hostname === "localhost" ||
+    // [::1] is the IPv6 localhost address.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:15]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    window.location.hostname === "[::1]" ||
+    // 127.0.0.0/8 are considered localhost for IPv4.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:17]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+    window.location.hostname.match(
+      /^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
+    )
+);
+
+export function register(config) {
+  if (process.env.NODE_ENV === "production" && "serviceWorker" in navigator) {
+    // The URL constructor is available in all browsers that support SW.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:25]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+    const publicUrl = new URL(process.env.PUBLIC_URL, window.location.href);
+    if (publicUrl.origin !== window.location.origin) {
+      // Our service worker won't work if PUBLIC_URL is on a different origin
+      // from what our page is served on. This might happen if a CDN is used to
+      // serve assets; see https://github.com/facebook/create-react-app/issues/2374
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:30]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+      return;
+    }
+
+    window.addEventListener("load", () => {
+      const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`;
+
+      if (isLocalhost) {
+        // This is running on localhost. Let's check if a service worker still exists or not.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:38]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        checkValidServiceWorker(swUrl, config);
+
+        // Add some additional logging to localhost, pointing developers to the
+        // service worker/PWA documentation.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:42]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+        navigator.serviceWorker.ready.then(() => {
+          console.log(
+            "This web app is being served cache-first by a service " +
+              "worker. To learn more, visit https://cra.link/PWA"
+          );
+        });
+      } else {
+        // Is not localhost. Just register service worker
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:50]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+        const installingWorker = registration.installing;
+        if (installingWorker == null) {
+          return;
+        }
+        installingWorker.onstatechange = () => {
+          if (installingWorker.state === "installed") {
+            if (navigator.serviceWorker.controller) {
+              // At this point, the updated precached content has been fetched,
+              // but the previous service worker will still serve the older
+              // content until all client tabs are closed.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:71]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+              console.log(
+                "New content is available and will be used when all " +
+                  "tabs for this page are closed. See https://cra.link/PWA."
+              );
+
+              // Execute callback
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:77]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+              if (config && config.onUpdate) {
+                config.onUpdate(registration);
+              }
+            } else {
+              // At this point, everything has been precached.
+              // It's the perfect time to display a
+              // "Content is cached for offline use." message.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:84]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+              console.log("Content is cached for offline use.");
+
+              // Execute callback
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:87]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+        };
+      };
+    })
+    .catch((error) => {
+      console.error("Error during service worker registration:", error);
+    });
+}
+
+function checkValidServiceWorker(swUrl, config) {
+  // Check if the service worker can be found. If it can't reload the page.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:102]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+  fetch(swUrl, {
+    headers: { "Service-Worker": "script" },
+  })
+    .then((response) => {
+      // Ensure service worker exists, and that we really are getting a JS file.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:107]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+      const contentType = response.headers.get("content-type");
+      if (
+        response.status === 404 ||
+        (contentType != null && contentType.indexOf("javascript") === -1)
+      ) {
+        // No service worker found. Probably a different app. Reload the page.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:113]: 🐘

File: client/src/serviceWorkerRegistration.js [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+        navigator.serviceWorker.ready.then((registration) => {
+          registration.unregister().then(() => {
+            window.location.reload();
+          });
+        });
+      } else {
+        // Service worker found. Proceed as normal.
```
Comment (arassaria) [inline client/src/serviceWorkerRegistration.js:120]: 🐘

**Review by arassaria — COMMENTED** [review]
🐘🐘🐘🐘🐘🐘🐘🐘🐘🐘
🐘🐘🐘🐘🐘🐘🐘🐘🐘🐘
🐘🐘🐘🐘🐘🐘🐘🐘🐘🐘
🐘

Why are there 31 comments in two files?

File: client/src/service-worker.js (hunk 1 above)
Comment (bwnstck) [reply client/src/service-worker.js:8]: for my understanding yes 👍

**Review by arassaria — APPROVED** [review]
LGTM 🚀 👨🏻‍🎤

**Pull request merged**